### PR TITLE
chore: enable `prefer-ast-types-enum` internal rule

### PR DESCRIPTION
**Pull request by G-Rath**
This is based off #1508, so shouldn't be merged until after that one has.

### CODE REVIEW

**Comment by G-Rath** [comment]
Looking over the diff, there are some pretty decent ones in there that have been caught & fixed.

`constructor` is the only "ehhh" one that looks like it could be considered overzealous.

File: packages/eslint-plugin/src/rules/member-naming.ts [hunk 1]
```diff
@@ -94,7 +94,7 @@ export default util.createRule<Options, MessageIds>({
     ): void {
       if (
         node.type === AST_NODE_TYPES.MethodDefinition &&
-        node.kind === 'constructor'
+        node.kind === AST_NODE_TYPES.constructor
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/member-naming.ts]: this is not correct

Comment (G-Rath) [reply packages/eslint-plugin/src/rules/member-naming.ts]: Yeah, I've got to push my latest re-lint.

It's because I forgot to run `yarn build` :)

**Comment by G-Rath** [comment]
@armano2 now this should be good & ready for review :)

**Comment by G-Rath** [comment]
👍 

> I should probably prioritise finishing some of those "abandoned" PRs.

I'm happy to help out with this - I've already got a couple on my plate that I'd like to action but they're all somewhat complicated or have high depth, so I'm keen for some lighter-weight contributions to sink my teeth into for upskilling 🙂

(In particular, learning how to use the types provided by TypeScript properly)

File: packages/eslint-plugin-internal/tests/rules/prefer-ast-types-enum.test.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+node.type === 'Literal'
+node.type === 'Keyword'
+    `,
+    output: `
+node.type === AST_NODE_TYPES.Literal
+node.type === AST_TOKEN_TYPES.Keyword
+    `,
+    errors: [
+      {
+        data: { enumName: 'AST_NODE_TYPES', literal: AST_NODE_TYPES.Literal },
```
Comment (G-Rath) [inline packages/eslint-plugin-internal/tests/rules/prefer-ast-types-enum.test.ts]: ah haha caught by my own rule!

File: packages/experimental-utils/src/ts-eslint/Scope.ts [hunk 3]
```diff
@@ -65,7 +65,11 @@ namespace Scope {
   }
 
   export type DefinitionType =
-    | { type: 'CatchClause'; node: TSESTree.CatchClause; parent: null }
+    | {
+        type: AST_NODE_TYPES.CatchClause;
```
Comment (armano2) [inline packages/experimental-utils/src/ts-eslint/Scope.ts]: False positive

File: packages/eslint-plugin/src/rules/no-unused-vars-experimental.ts [hunk 4]
```diff
@@ -160,7 +161,7 @@ export default util.createRule<Options, MessageIds>({
           break;
 
         case ts.SyntaxKind.PropertyDeclaration:
-          report('Property');
+          report(AST_NODE_TYPES.Property);
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/no-unused-vars-experimental.ts]: False positive

File: packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  AST_TOKEN_TYPES,
+  ESLintUtils,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+const isStringLiteral = (
+  node: TSESTree.Literal,
+): node is TSESTree.StringLiteral => typeof node.value === 'string';
+
+export = ESLintUtils.RuleCreator(name => name)({
```
Comment (armano2) [inline packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts]:
```suggestion
export default createRule({
```

File: packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import {
+  AST_NODE_TYPES,
+  AST_TOKEN_TYPES,
+  ESLintUtils,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
```
Comment (armano2) [inline packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts]:
```suggestion
import { createRule } from '../util';

```

File: packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
+import {
+  AST_NODE_TYPES,
+  AST_TOKEN_TYPES,
+  ESLintUtils,
```
Comment (armano2) [inline packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts]:
```suggestion
```

**Comment by armano2** [comment]
bdw. what do you think about changing this validation to something like this

```ts
function checkNodes(
  node: TSESTree.Node,
  literal: TSESTree.StringLiteral,
): boolean {
  switch (node.type) {
    case AST_NODE_TYPES.MemberExpression:
    case AST_NODE_TYPES.OptionalMemberExpression:
      return checkNodes(node.property, literal);
    case AST_NODE_TYPES.BinaryExpression:
      return node.right === literal ? checkNodes(node.left, literal) : false;
    case AST_NODE_TYPES.Property:
      return node.value === literal ? checkNodes(node.key, literal) : false;
    case AST_NODE_TYPES.SwitchCase:
      return node.test === literal;
    case AST_NODE_TYPES.Identifier:
      return node.name === 'type' || node.name === 'node';
    case AST_NODE_TYPES.ArrayExpression:
    case AST_NODE_TYPES.TSTypeReference:
      return true;
  }
  return false;
}

...
        if (
          !isStringLiteral(node) ||
          !node.parent ||
          !checkNodes(node.parent, node)
        ) {
          return;
        }
```

this should eliminate most of false positives

**Comment by G-Rath** [comment]
> what do you think about changing this validation to something like this

I think that while cool it's probably worth focusing our efforts on getting #315 across the line before iterating any further on this rule, as I think we've got the majority of the false false positives, and it'll also make for a good double-blind test for that rule :)

**Pull request merged**